src/IO/StdStreamFromReadBuffer.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include <IO/StdStreamBufFromReadBuffer.h>
+#include <iostream>
+#include <memory>
+
+
+namespace DB
+{
+class ReadBuffer;
+
+/// `std::istream`-compatible wrapper around a ReadBuffer.
+class StdIStreamFromReadBuffer : public std::istream
+{
+public:
+    using Base = std::istream;
+    StdIStreamFromReadBuffer(std::unique_ptr<ReadBuffer> buf, size_t size) : Base(&stream_buf), stream_buf(std::move(buf), size) { }
+    StdIStreamFromReadBuffer(ReadBuffer & buf, size_t size) : Base(&stream_buf), stream_buf(buf, size) { }
+    StdStreamBufFromReadBuffer * rdbuf() const { return const_cast<StdStreamBufFromReadBuffer *>(&stream_buf); }
+
+private:
+    StdStreamBufFromReadBuffer stream_buf;
+};
+
+
+/// `std::iostream`-compatible wrapper around a ReadBuffer.
+class StdStreamFromReadBuffer : public std::iostream
+{
+public:
+    using Base = std::iostream;
+    StdStreamFromReadBuffer(std::unique_ptr<ReadBuffer> buf, size_t size) : Base(&stream_buf), stream_buf(std::move(buf), size) { }
+    StdStreamFromReadBuffer(ReadBuffer & buf, size_t size) : Base(&stream_buf), stream_buf(buf, size) { }
```

Comment:
I have some questions about this, 

`Base(&stream_buf)` will call the following constructor

```cpp
template <class _CharT, class _Traits>
basic_streambuf<_CharT, _Traits>::basic_streambuf()
    : __binp_(nullptr),
      __ninp_(nullptr),
      __einp_(nullptr),
      __bout_(nullptr),
      __nout_(nullptr),
      __eout_(nullptr)
{
}

template <class _CharT, class _Traits>
basic_streambuf<_CharT, _Traits>::basic_streambuf(const basic_streambuf& __sb)
    : __loc_(__sb.__loc_),
      __binp_(__sb.__binp_),
      __ninp_(__sb.__ninp_),
      __einp_(__sb.__einp_),
      __bout_(__sb.__bout_),
      __nout_(__sb.__nout_),
      __eout_(__sb.__eout_)
{
}
```

so, `__ninp_` and `__einp_` are nullptr. But std::iostream will call `sgetc()` to get char, so we will always call `underflow()` which does not advance the position.

```cpp
inline _LIBCPP_HIDE_FROM_ABI_AFTER_V1
int_type sgetc() {
    if (__ninp_ == __einp_)
        return underflow();
    return traits_type::to_int_type(*__ninp_);
}
```